### PR TITLE
HPCC-21949 configmgr wizard does not add stock keys section

### PR DESCRIPTION
**Pull request by kenrowland**
Updated envmod cli to support adding keys section

Signed-off-by: Ken Rowland kenneth.rowland@lexisnexisrisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: configuration/cli/envmod/envmod.cpp [hunk 1]
```diff
@@ -48,8 +48,8 @@ std::vector<std::string> splitString(const std::string &input, const std::string
 // Default configuration directories
 EnvironmentType envType = XML;
 std::string masterSchemaFile = "environment.xsd";
-std::string configSchemaDir = "";
-std::string configSchemaRelativeDir = COMPONENTFILES_DIR PATHSEPSTR "configschema" PATHSEPSTR "xsd" PATHSEPSTR;
+//std::string configSchemaDir = "";
```
Comment (rpastrana) [inline configuration/cli/envmod/envmod.cpp:51]: remove

File: configuration/cli/envmod/envmod.cpp [hunk 2]
```diff
@@ -48,8 +48,8 @@ std::vector<std::string> splitString(const std::string &input, const std::string
...
+std::string configSchemaDir = COMPONENTFILES_DIR PATHSEPSTR "configschema" PATHSEPSTR "xsd" PATHSEPSTR;
```
Comment (rpastrana) [inline configuration/cli/envmod/envmod.cpp:52]: this seems to only be used in the usage output, is it actually needed?

File: configuration/cli/envmod/envmod.cpp [hunk 3]
```diff
@@ -118,7 +118,7 @@ int main(int argc, char *argv[])
     //
     // Build the default directory for the schema files
     std::string processPath(queryCurrentProcessPath());
-    configSchemaDir = processPath.substr(0, processPath.find_last_of(PATHSEPSTR)) + PATHSEPSTR + configSchemaRelativeDir;
+    //configSchemaDir = processPath.substr(0, processPath.find_last_of(PATHSEPSTR)) + PATHSEPSTR + configSchemaRelativeDir;
```
Comment (rpastrana) [inline configuration/cli/envmod/envmod.cpp:121]: remove

File: configuration/cli/envmod/envmod.cpp [hunk 4]
```diff
@@ -118,7 +118,7 @@ int main(int argc, char *argv[])
     //
     // Build the default directory for the schema files
```
Comment (rpastrana) [inline configuration/cli/envmod/envmod.cpp:119]: does this comment still apply?

File: initfiles/componentfiles/configschema/templates/AddKeys.json [hunk 5]
```diff
@@ -1,72 +1,129 @@
...
             "name" : "keypair_name",
             "type" : "string",
-            "value" : "mythor"
+            "values" : ["mythor"]
         },
         {
             "name" : "public_key",
             "type" : "string",
-            "value" : "/home/jsmith/.ssh/id_rsa.pub.pem"
+            "values" : ["/home/jsmith/.ssh/id_rsa.pub.pem"]
```
Comment (rpastrana) [inline initfiles/componentfiles/configschema/templates/AddKeys.json:13]: are the contents of "values" serving as default values?
Is the content of the values attribute actually inserted into the target env file?
I was expecting place holders, but in this particular case the path seems to be exemplary only.
If its not a placeholder, and we can't use build time vars from the platform to construct these values, let's try to match the out of the box env.xml:
@HOME_DIR@/@RUNTIME_USER@/certificate/public.key.pem
how about "/home/your_user/certificate/plublic.key.pem"

**Review by rpastrana — COMMENTED** [review]
@kenrowland a few comments

File: configuration/cli/envmod/envmod.cpp (hunk 4 above)
Comment (kenrowland) [reply configuration/cli/envmod/envmod.cpp]: It does not.

File: configuration/cli/envmod/envmod.cpp (hunk 3 above)
Comment (kenrowland) [reply configuration/cli/envmod/envmod.cpp]: Gone.

File: initfiles/componentfiles/configschema/templates/AddKeys.json (hunk 5 above)
Comment (kenrowland) [reply initfiles/componentfiles/configschema/templates/AddKeys.json:13]: This is an example modification template that adds keys. These are example values that would be replaced. I left the template in the PR (actually it has been in the repo for a while) as a discussion point for where these go. Should they be in the repo at all. I feel the decision is outside my realm. So, I expect once the decision is made, we may modify this like what you did above, or maybe just add <place your path here>.

**Review by kenrowland — COMMENTED** [review]
@rpastrana Please changes and comments, especially about the AddKeys.json file

File: configuration/cli/envmod/envmod.cpp (hunk 2 above)
Comment (rpastrana) [reply configuration/cli/envmod/envmod.cpp:51]: configSchemaDir seems to only be referenced in the usage, and doesn't seem to be reassigned, it should at least be const?

File: initfiles/componentfiles/configschema/templates/AddKeys.json (hunk 5 above)
Comment (rpastrana) [reply initfiles/componentfiles/configschema/templates/AddKeys.json:13]: Per offline discussion, this file will be removed and documented elsewhere

**Review by rpastrana — COMMENTED** [review]
@kenrowland one comment

File: configuration/cli/envmod/envmod.cpp (hunk 2 above)
Comment (kenrowland) [reply configuration/cli/envmod/envmod.cpp:51]: configSchemaDir is settable with a parameter, so can't be const.

File: initfiles/componentfiles/configschema/templates/AddKeys.json (hunk 5 above)
Comment (kenrowland) [reply initfiles/componentfiles/configschema/templates/AddKeys.json]: Spoke to Stu, he leans toward leaving it in the repo for now. I am updating the values to show they need to be changed. Richard does not have a preference, so for now just going to leave it in the repo.

**Review by kenrowland — COMMENTED** [review]
@rpastrana Please review latest changes.

**Review by kenrowland — COMMENTED** [review]
@rpastrana Please see my final comments.

**Review by rpastrana — APPROVED** [review]

**Comment by kenrowland** [comment]
@richardkchapman Please merge. Thanks.

**Comment by kenrowland** [comment]
@richardkchapman Please merge (forgot to rebase last time I asked)

File: configuration/configmgr/configmgrlib/mod_template_support/Variables.cpp [hunk 6]
```diff
@@ -90,16 +90,20 @@ std::string Variables::doValueSubstitution(const std::string &value) const
     std::size_t bracesStartPos = result.find("{{");
     bool done = bracesStartPos == std::string::npos;
 
+    // {{hwips[{{numsystems}}-{{ccnum}}]}}
+    // ^                                ^
+    // 0                                33
```
Comment (jakesmith) [inline configuration/configmgr/configmgrlib/mod_template_support/Variables.cpp:95]: was this comment supposed to be left in?

File: configuration/configmgr/configmgrlib/mod_template_support/Variables.cpp [hunk 7]
```diff
@@ -90,16 +90,20 @@ std::string Variables::doValueSubstitution(const std::string &value) const
...
         std::size_t bracesEndPos = findClosingDelimiter(result, bracesStartPos,"{{", "}}");
-        varName = result.substr(bracesStartPos + 2, bracesEndPos - bracesStartPos - 2);
+        varName = result.substr(bracesStartPos + 2, bracesEndPos - bracesStartPos - 2);  // hwips[{{numsystems}}-{{ccnum}}]
 
         //
         // If there is an index defined, evaluate it and update the index to be used for the final value
-        std::size_t bracketStartPos = result.find('[');
-        std::size_t sizePos = result.find(".size");
+        std::size_t bracketStartPos = varName.find('[');  // 5
+        std::size_t sizePos = varName.find(".size");  // npos
```
Comment (jakesmith) [inline configuration/configmgr/configmgrlib/mod_template_support/Variables.cpp:106]: trivial: not sure those comments are helpful/meaningful to other reads

**Review by jakesmith — COMMENTED** [review]
Looks okay afaics, a few minor comments.

File: configuration/cli/envmod/envmod.cpp [hunk 8]
```diff
@@ -48,8 +48,7 @@ std::vector<std::string> splitString(const std::string &input, const std::string
 // Default configuration directories
 EnvironmentType envType = XML;
 std::string masterSchemaFile = "environment.xsd";
-std::string configSchemaDir = "";
-std::string configSchemaRelativeDir = COMPONENTFILES_DIR PATHSEPSTR "configschema" PATHSEPSTR "xsd" PATHSEPSTR;
+std::string configSchemaDir = COMPONENTFILES_DIR PATHSEPSTR "configschema" PATHSEPSTR "xsd" PATHSEPSTR;
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/cli/envmod/envmod.cpp:51]: From what you say, there are no required changes here, but we may need to do something when up merged. I looked at the referenced Jira and associated PR. The changes for a custom merge, if needed, should be minimal.

File: configuration/cli/envmod/envmod.cpp [hunk 9]
```diff
@@ -115,10 +114,7 @@ void usage()
 
 int main(int argc, char *argv[])
 {
-    //
-    // Build the default directory for the schema files
     std::string processPath(queryCurrentProcessPath());
-    configSchemaDir = processPath.substr(0, processPath.find_last_of(PATHSEPSTR)) + PATHSEPSTR + configSchemaRelativeDir;
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/cli/envmod/envmod.cpp:121]: Let's wait for the up-merge to see what is fully required.

File: configuration/configmgr/configmgrlib/EnvironmentNode.cpp [hunk 10]
```diff
@@ -319,7 +319,7 @@ void EnvironmentNode::getInsertableItems(std::vector<InsertableItem> &insertable
         }
         else
         {
-            childCounts.insert({ itemType, 1 });
+            childCounts.insert({ itemType, 0 });
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/configmgr/configmgrlib/EnvironmentNode.cpp:322]: Directly related.

File: configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp [hunk 11]
```diff
@@ -312,7 +312,7 @@ void EnvModTemplate::parseOperation(const rapidjson::Value &operation)
     {
         pOp = std::make_shared<OperationFindNode>();
     }
-    else if (action == "delete_")
+    else if (action == "delete")
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:315]: Directly related.

File: configuration/configmgr/configmgrlib/mod_template_support/Variables.cpp (hunk 6 above)
Comment (kenrowland) [reply configuration/configmgr/configmgrlib/mod_template_support/Variables.cpp]: Was a debugging aid when stepping through the code. Removed.

File: configuration/configmgr/configmgrlib/mod_template_support/Variables.cpp (hunk 7 above)
Comment (kenrowland) [reply configuration/configmgr/configmgrlib/mod_template_support/Variables.cpp]: Again, left over debugging comments. Removed along with some others I found.

**Review by kenrowland — COMMENTED** [review]
@jakesmith Updated from your comments. Please review.

Also, in case you were not aware, this is a tool that is used to modify the environment.xml for cases where the config manager cannot easily support required config changes.

**Review by jakesmith — APPROVED** [review]
Looks good, please squash

**Comment by kenrowland** [comment]
@richardkchapman Please merge

**Pull request merged**